lib/hanrahashiv.py

```diff
@@ -376,6 +375,164 @@ def ReplaceItemsInFormula(
+            for k in range(len(new_replace_items[j])):
+                new_replace_items[j][k] += len(replace_items_strs[i]) - len(new_replace_items[i])
+        for j in range(len(added_items)):
+            for k in range(len(added_items[j])):
+                if added_items[j][k] > item_index:
+                    added_items[j][k] += len(replace_items_strs[i]) - len(new_replace_items[i])
+        replaced_items.append(list(range(item_index, item_index + len(replace_items_strs[i]))))
+
+    flattened_replaced_items = [item for sublist in replaced_items for item in sublist]
+    flattened_added_items = [item for sublist in added_items for item in sublist]
+    flattened_replace_items = [item for sublist in replace_items for item in sublist]
+
+    # find indices of fixed item to transformed later
+    fixed_items_old = []
+    for i in range(len(tex_string_list)):
+        if i not in remove_items and i not in flattened_replace_items:
+            fixed_items_old.append(i)
+    fixed_items_new = []
+    for i in range(len(new_tex_string_list)):
+        if i not in flattened_added_items and i not in flattened_replaced_items:
+            fixed_items_new.append(i)
+
+    new_formula = type(formula)(
+        *new_tex_string_list,
+        font_size=formula[0].font_size, tex_template=formula.tex_template
+    )
+    new_formula.move_to(formula).align_to(formula, new_formula_alignment)
+
+    # fix colors
+    for i in range(len(fixed_items_old)):
+        new_formula[fixed_items_new[i]].set_color(formula[fixed_items_old[i]].color)
+    for i in range(len(add_items_colors)):
+        for j in range(len(add_items_colors[i])):
+            if add_items_colors[i][j]:
+                new_formula[added_items[i][j]].set_color(add_items_colors[i][j])
+    for i in range(len(replace_items_strs)):
+        for j in range(len(replace_items_strs[i])):
+            new_formula[replaced_items[i][j]].set_color(formula[replace_items[i][0]].color)
+    for i in range(len(replace_items_colors)):
+        for j in range(len(replace_items_colors[i])):
+            if replace_items_colors[i][j]:
+                new_formula[replaced_items[i][j]].set_color(replace_items_colors[i][j])
+
+    # animation groups
+    animations = []
+    animations += [FadeOut(formula[i]) for i in remove_items]
+    animations += [
+        ReplacementTransform(formula[replace_items[i][0] : replace_items[i][-1] + 1],
+        new_formula[replaced_items[i][0] : replaced_items[i][-1] + 1])
+        for i in range(len(replace_items))
+    ]
+    animations += [
+        ReplacementTransform(formula[fixed_items_old[i]], new_formula[fixed_items_new[i]])
+        for i in range(len(fixed_items_new))
+    ]
+    animations_add_items = [add_items_animation_style(new_formula[i]) for i in flattened_added_items]
+
+    formula.remove(*formula)
+    formula.add(*new_formula)
+
+    return AnimationGroup(AnimationGroup(*animations), AnimationGroup(*animations_add_items), lag_ratio=add_lag_ratio)

```

Comment:
Did you do any other tests besides `formula_modifications.py`? I think this file needs a little bit more testing